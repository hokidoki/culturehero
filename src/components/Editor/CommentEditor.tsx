import React, { useCallback, useEffect, useRef, useState } from 'react';
import ContentEditable from 'react-contenteditable';
import CommentEditorStyles from './CommentEditor.styles';
import { CommentEditorProps } from './CommentEditor.types';
import { editorClickOutside } from '../utils';

export default function CommentEditor({
  article: __article__ = '',
  disabled,
  onSubmit,
}: CommentEditorProps) {
  const ref = useRef<HTMLDivElement>(null);
  const [article, setArticle] = useState(__article__);

  const submitWrap = useCallback(
    (comment: string) => {
      onSubmit(comment)
        .then(() => setArticle(''))
        .catch(() => {});
    },
    [onSubmit]
  );

  return (
    <CommentEditorStyles.Layout>
      <CommentEditorStyles.CommentTextArea
        onClick={(e) => editorClickOutside(e, ref.current as HTMLDivElement)}
      >
        <ContentEditable
          innerRef={ref}
          html={article}
          disabled={disabled}
          onChange={(e) => setArticle(e.target.value)}
        />
      </CommentEditorStyles.CommentTextArea>
      <CommentEditorStyles.SubmitButton
        disabled={disabled}
        onClick={() => submitWrap(article)}
      >
        쓰기
      </CommentEditorStyles.SubmitButton>
    </CommentEditorStyles.Layout>
  );
}
