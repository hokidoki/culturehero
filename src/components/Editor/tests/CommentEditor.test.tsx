import { fireEvent, render, waitFor } from '@render';
import CommentEditor from '../CommentEditor';
import { isNull } from '@fxts/core';

describe('CommentEditor', () => {
  it('match snapshot', () => {
    const props = {
      article: 'ARTICLE',
      onSubmit: () => Promise.resolve(),
      disabled: false,
    };
    const utils = render(<CommentEditor {...props} />);
    expect(utils.container).toMatchSnapshot();
  });

  it('onSubmit이 resolved 상태라면, [contenteditable]요소의 내용은 삭제된다.', async () => {
    const props = {
      article: 'ARTICLE',
      onSubmit: jest.fn().mockResolvedValue('SUCCESS'),
      disabled: false,
    };
    const utils = render(<CommentEditor {...props} />);
    const editor = utils.container.querySelector('[contenteditable]');
    const button = utils.container.querySelector('button');

    if (isNull(button) || isNull(editor)) throw new Error('Element is null');

    expect(editor.innerHTML).toBe(props.article);

    fireEvent.click(button);

    await waitFor(() => expect(props.onSubmit).toBeCalled());

    expect(editor.innerHTML).toBe('');
  });

  it('onSubmit이 rejected상태라면, [contenteditable]요소의 내용은 이전과 동일하다.', async () => {
    const props = {
      article: 'ARTICLE',
      onSubmit: jest.fn().mockRejectedValue('FAILED'),
      disabled: false,
    };
    const utils = render(<CommentEditor {...props} />);
    const editor = utils.container.querySelector('[contenteditable]');
    const button = utils.container.querySelector('button');

    if (isNull(button) || isNull(editor)) throw new Error('Element is null');

    expect(editor.innerHTML).toBe(props.article);

    fireEvent.click(button);

    await waitFor(() => expect(props.onSubmit).toBeCalled());

    expect(editor.innerHTML).toBe(props.article);
  });
});
