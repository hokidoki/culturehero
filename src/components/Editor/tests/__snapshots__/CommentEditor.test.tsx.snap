// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentEditor match snapshot 1`] = `
.c0 {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: flex-start;
  width: 100%;
  height: 100%;
  gap: 20px;
  flex-grow: 1;
  padding: 20px;
}

.c1 {
  width: 100%;
  height: 100%;
  border-radius: 10px;
  background-color: transparent;
  outline: none;
  resize: none;
  padding: 10px;
  font-size: 1rem;
  color: #9aa0a6;
  background-color: rgb(33,33,36);
  cursor: text;
}

.c1 div[contenteditable=true] {
  outline: none;
}

.c2 {
  width: 100px;
  height: 100%;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
  background-color: transparent;
  font-size: 1.25rem;
  color: white;
  transition: all 0.5s;
}

.c2:disabled {
  color: #bdc1c6;
  cursor: not-allowed;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        contenteditable="true"
      >
        ARTICLE
      </div>
    </div>
    <button
      class="c2"
    >
      쓰기
    </button>
  </div>
</div>
`;
